velox/exec/HashBuild.cpp

```diff
@@ -178,12 +278,315 @@ void HashBuild::addInput(RowVectorPtr input) {
+bool HashBuild::reserveMemory(const RowVectorPtr& input) {
+  VELOX_CHECK(spillEnabled());
+
+  numSpillRows_ = 0;
+  numSpillBytes_ = 0;
+
+  auto rows = table_->rows();
+  auto numRows = rows->numRows();
+  if (numRows == 0) {
+    // Build side is empty. Nothing to spill.
+    return true;
+  }
+
+  // TODO: add table size and increment estimation.
+
+  auto [freeRows, outOfLineFreeBytes] = rows->freeSpace();
+  auto outOfLineBytes =
+      rows->stringAllocator().retainedSize() - outOfLineFreeBytes;
+  auto outOfLineBytesPerRow = std::max<uint64_t>(1, outOfLineBytes / numRows);
+  int64_t flatBytes = input->estimateFlatSize();
+
+  // Test-only spill path.
+  if (testingTriggerSpill()) {
+    numSpillRows_ = std::max<int64_t>(1, numRows / 10);
+    numSpillBytes_ = numSpillRows_ * outOfLineBytesPerRow;
+    return false;
+  }
+
+  if (freeRows > input->size() &&
+      (outOfLineBytes == 0 || outOfLineFreeBytes >= flatBytes)) {
+    // Enough free rows for input rows and enough variable length free
+    // space for the flat size of the whole vector. If outOfLineBytes
+    // is 0 there is no need for variable length space.
+    return true;
+  }
+
+  // If there is variable length data we take the flat size of the
+  // input as a cap on the new variable length data needed.
+  const auto increment =
+      rows->sizeIncrement(input->size(), outOfLineBytes ? flatBytes : 0);
+
+  auto tracker = CHECK_NOTNULL(mappedMemory_->tracker());
+  // There must be at least 2x the increments in reservation.
+  if (tracker->getAvailableReservation() > 2 * increment) {
+    return true;
+  }
+
+  // Check if we can increase reservation. The increment is the larger of
+  // twice the maximum increment from this input and
+  // 'spillableReservationGrowthPct_' of the current reservation.
+  auto targetIncrement = std::max<int64_t>(
+      increment * 2,
+      tracker->getCurrentUserBytes() *
+          spillConfig()->spillableReservationGrowthPct / 100);
+  if (tracker->maybeReserve(targetIncrement)) {
+    return true;
+  }
+  numSpillRows_ = std::max<int64_t>(
+      1, targetIncrement / (rows->fixedRowSize() + outOfLineBytesPerRow));
+  numSpillBytes_ = numSpillRows_ * outOfLineBytesPerRow;
+  return false;
+}
+
+void HashBuild::spillInput(RowVectorPtr& input) {
+  VELOX_CHECK_EQ(input->size(), activeRows_.size());
+
+  if (!spillEnabled() || !spiller_->isAnySpilled() ||
+      !activeRows_.hasSelections()) {
+    return;
+  }
+
+  const auto numInput = input->size();
+  prepareInputIndicesBuffers(numInput, spiller_->spilledPartitionSet());
+  computeSpillPartitions(input);
+
+  vector_size_t numSpillInputs = 0;
+  for (auto row = 0; row < numInput; ++row) {
+    const auto partition = spillPartitions_[row];
+    if (FOLLY_UNLIKELY(!activeRows_.isValid(row))) {
+      continue;
+    }
+    if (!spiller_->isSpilled(partition)) {
+      continue;
+    }
+    activeRows_.setValid(row, false);
+    ++numSpillInputs;
+    spillInputIndicesBuffers_[partition]
```

Comment:
Would it make sense to cache raw pointers to the buffers to avoid the cast for each row?

Reply:
Yeah, good point. Updated!